go-controller/pkg/clustermanager/egressip_controller.go

```diff
@@ -0,0 +1,1768 @@
+			// even though statusToRemove will be removed afterwards
+			eIPC.deleteAllocatorEgressIPAssignments(statusToRemove)
+		}
+		if len(ipsToAssign) > 0 {
+			statusToAdd = eIPC.assignEgressIPs(name, ipsToAssign.UnsortedList())
+			statusToKeep = append(statusToKeep, statusToAdd...)
+		}
+		// Add all assignments which are to be kept to the allocator cache,
+		// allowing us to track all assignments which have been performed and
+		// avoid incorrect future assignments due to a de-synchronized cache.
+		eIPC.addAllocatorEgressIPAssignments(name, statusToKeep)
+		// Update the object only on an ADD/UPDATE. If we are processing a
+		// DELETE, new will be nil and we should not update the object.
+		if len(statusToAdd) > 0 || (len(statusToRemove) > 0 && new != nil) {
+			if err := eIPC.patchReplaceEgressIPStatus(name, statusToKeep); err != nil {
+				return err
+			}
+		}
+	} else {
+		// Even when running on a public cloud, we must make sure that we unwire EgressIP
+		// configuration from OVN *before* we instruct the CloudNetworkConfigController
+		// to remove the CloudPrivateIPConfig object from the cloud.
+		// CloudPrivateIPConfig objects can be in the "Deleting" state for a long time,
+		// waiting for the underlying cloud to finish its action and to report success of the
+		// unattach operation. Some clouds such as Azure will remove the IP address nearly
+		// immediately, but then they will take a long time (seconds to minutes) to actually report
+		// success of the removal operation.
+		if len(statusToRemove) > 0 {
+			// Delete all assignments that are to be removed from the allocator
+			// cache. If we don't do this we will occupy assignment positions for
+			// the ipsToAdd, even though statusToRemove will be removed afterwards
+			eIPC.deleteAllocatorEgressIPAssignments(statusToRemove)
+			// Before updating the cloud private IP object, we need to remove the OVN configuration
+			// for these invalid statuses so that traffic is not blackholed to non-existing setup in the
+			// cloud. Thus we patch the egressIP status with the valid set of statuses which will
+			// trigger an event for the ovnkube-master to take action upon.
+			// Note that once we figure out the statusToAdd parts below we will trigger an
+			// update to cloudPrivateIP object which will trigger another patch for the eIP object.
+			if err := eIPC.patchReplaceEgressIPStatus(name, statusToKeep); err != nil {
+				return err
+			}
+		}
+		// When egress IP is not fully assigned to a node, then statusToRemove may not
+		// have those entries, hence retrieve it from staleEgressIPs for removing
+		// the item from cloudprivateipconfig.
+		for _, toRemove := range statusToRemove {
+			if !staleEgressIPs.Has(toRemove.EgressIP) {
+				continue
+			}
+			staleEgressIPs.Delete(toRemove.EgressIP)
+		}
+		for staleEgressIP := range staleEgressIPs {
+			if nodeName := eIPC.deleteAllocatorEgressIPAssignmentIfExists(name, staleEgressIP); nodeName != "" {
```

Comment:
@kyrtapz and as for this case, I checked on master code, it seems we weren't called delete setup for NBDB for the stale ones so I don't think I need to do any patch replace here...